From def264ecf556784035ddc48c1ba67b2c327f8a79 Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlosg@gnome.org>
Date: Wed, 1 Jul 2026 10:32:51 +0200
Subject: [PATCH] backends/native: Ensure GPU data when asking for gbm_device

If the native cursor renderer is created at a time that a MetaGpuKms got
its MetaRendererNative's GPU data collected by free_unused_gpu_datas(), it
will be asking for a gbm_device in order to figure out whether using the
cursor plane is an option, but triggering a crash since the missing GPU data
would cause a NULL pointer dereference.

Since the cursor renderer is asking for a gbm_device, it makes sense to
make the MetaRendererNative get out of its way to get one. Ensure to create
GPU data for it, or sanely bail out if it could not be created on the spot.

Closes: https://gitlab.gnome.org/GNOME/mutter/-/work_items/4890
Fixes: de5ee29e94 ("backends/native: Initialize HW cursor support right away")
Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/5153>
---
 src/backends/native/meta-renderer-native.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/backends/native/meta-renderer-native.c b/src/backends/native/meta-renderer-native.c
index f11ae8bdb4b..6b5c8deefed 100644
--- a/src/backends/native/meta-renderer-native.c
+++ b/src/backends/native/meta-renderer-native.c
@@ -174,9 +174,18 @@ meta_gbm_device_from_gpu (MetaGpuKms *gpu_kms)
   MetaRendererNativeGpuData *renderer_gpu_data;
   MetaRenderDevice *render_device;
   MetaRenderDeviceGbm *render_device_gbm;
+  g_autoptr (GError) error = NULL;
+
+  if (!meta_renderer_native_ensure_gpu_data (renderer_native, gpu_kms, &error))
+    {
+      g_warning ("meta_gbm_device_from_gpu: could not create gpu_data for gpu %s: %s",
+                 meta_gpu_kms_get_file_path (gpu_kms), error->message);
+      return NULL;
+    }
 
   renderer_gpu_data = meta_renderer_native_get_gpu_data (renderer_native,
                                                          gpu_kms);
+  g_assert (renderer_gpu_data);
   render_device = renderer_gpu_data->render_device;
 
   if (!META_IS_RENDER_DEVICE_GBM (render_device))
-- 
GitLab
